### ISSUE TITLE
Support items generated by proc macros

### ISSUE THREAD

**Issue by shssoichiro** [intellij-rust/intellij-rust#1786]
For example, let's say I am using the crate `derive_builder` with the following code snippet:

```rust
#[derive(Clone, Debug, Builder)]
pub struct MyStruct {
    foo: String
}
```

`derive_builder` generates a new struct named `MyStructBuilder`. I'd like to have this generated `MyStructBuilder` struct available as an autocomplete option to make its usage easier.

**Comment by farodin91** [comment]
Intellij-rust support `proc_macro_derive` only partial.

**Comment by natepisarski** [comment]
I'm seconding this one. This would be a real life saver with diesel. 

Correct me if I'm wrong, but it seems like full `proc_macro` support would be impossible. In diesel's case, an AST isn't enough to autcomplete table names. It seems like the case here is custom completions based on up-front knowlege about what the given macro will generate.

I could be totally off-base. It just seems like this would wind up being "add Builder support" moreso than all auto-generated struct support.

**Comment by shssoichiro** [comment]
I feel like implementing this for any proc macros would give us the same support for `diesel`. If there is a limitation, it is likely with diesel's `infer_schema`, since it requires connecting to the database to determine available table names. Users who use `diesel print-schema` from the CLI to generate their `schema.rs` file should benefit from this feature, or at least that would be my initial assumption.

**Comment by NyxCode** [comment]
Hell ya, we need this! It wouldn't work with `infer_schema`, but as far as I understood it it's only supposed to be used for testing or rapid prototyping - so it's not a big issue!

**Comment by electricjones** [comment]
Forgive my ignorance about how the inner guts of the plugin works, but would https://github.com/dtolnay/cargo-expand be useful for indexing generated macros? It's a cargo command that expands the macros out.

I've also found that RLS is able to see generated macros. VSCode and others offer autocomplete for derive-builder and others. Is there anyway to use RLS in this one instance?

**Comment by electricjones** [comment]
Actually, looking through recent activity @vlad20012, will #3015 resolve this? It looks like the same basic problem as #2458

**Comment by mchernyavsky** [comment]
> Actually, looking through recent activity @vlad20012, will #3015 resolve this? It looks like the same basic problem as #2458

No, because it doesn't handle proc macros (like `derive`).

> Forgive my ignorance about how the inner guts of the plugin works, but would https://github.com/dtolnay/cargo-expand be useful for indexing generated macros? It's a cargo command that expands the macros out.

`cargo-expand` has some drawbacks, the details you can find out from @fedochet. He is currently working on an [alternative solution](https://github.com/fedochet/rust-proc-macro-expander) that will be used for indexing proc macros.

P.S. - On nightly version of plugin we have simple integration with `cargo expand` (https://github.com/intellij-rust/intellij-rust/pull/3736):

<img width="400" alt="expand" src="https://user-images.githubusercontent.com/6079006/57088615-50d95900-6d0b-11e9-9ca3-f361614e97d9.png">

**Comment by fedochet** [comment]
@electricjones can you please provide some details about proc macro support by RLS and VSCode? 

I am not exactly familiar with this setup, so maybe I'm doing something wrong: I've just installed RLS and VSCode extension (as explained [here](https://github.com/rust-lang/rls-vscode#quick-start)), created new cargo project with `derive_builder` as a dependency. I took the sample code from `derive_builder` repo.

VSCode does not see `ChannelBuilder`:

![image](https://user-images.githubusercontent.com/8393617/57091532-23dc7480-6d12-11e9-9e47-3a1d0221febc.png)

Also it does not see any method in it: 

![image](https://user-images.githubusercontent.com/8393617/57091693-6aca6a00-6d12-11e9-82f3-6da2ec6143db.png)

There are `special_info` and `token` among the autocompletion list, but my guess is that VSCode proposes any token that is present in the current file (maybe I'm wrong)

Maybe there is something missing in my setup that I could fix?

**Comment by electricjones** [comment]
@fedochet, No I think you are right. My setup is behaving the same way.

When I use the `ChannelBuilder` like above I also see `special_info` and `token` and other fields I've added.

But, when I choose `token` from the list and add parentheses (turning it into a method call), VSCode *does* seem to recognize information about it. I get an error when I don't pass in a value
```
this function takes 1 parameter but 0 parameters were supplied

expected 1 parameterrustc(E0061)
main.rs(4, 19): defined here
```
The `main.rs(4, 19)` is my `derive(Builder)` declaration.

I get a similar error when trying to pass in an incorrect type.

 When I "go do definition" for any of those builder methods it takes me to the struct's definition of that field. 

I guess I just assumed all of that was intentional, and that RLS / VSCode handled procedural macros. Maybe it's just a combination of luck that it is working as it is.

**Comment by atroche** [comment]
I probably should've checked if you're okay with my doing this, but I've created a [bounty](https://www.bountysource.com/issues/49744518-add-generated-structs-as-autocomplete-suggestions) for this issue on BountySource to encourage contributions.

**Comment by exi** [comment]
This also affects some rust core library structs.

For example the output of "my,test".split(',') cannot be autocompleted/typed because the returned "Split" type is macro generated.

**Comment by Undin** [comment]
@exi note, your issue is not related to the original one. This issue is about items generated by [procedural](https://doc.rust-lang.org/reference/procedural-macros.html) macros.
In your case, `Split` struct is generated by declarative macros and everything should be fine with new macro expansion engine (see #3628)

**Comment by FelixBrakel** [comment]
Is there a workaround for this? I can see that the Serialize and Deserialize traits generated by the serde crate do get picked up by intellij-rust. Can I write my derive macros in such a way that they'll be picked up as well? Or is serde some kind of special case?

**Comment by Undin** [comment]
> Is there a workaround for this?

Unfortunately, not now.

> Or is serde some kind of special case?

Yep. We hardcoded `Serialize` and `Deserialize` support as the most commonly used procedural macros

**Comment by Restioson** [comment]
I wonder whether [watt](https://github.com/dtolnay/watt) would be in any way helpful here? It solves one possible issue: enforcing determinism in proc-macros. Maybe it could be used to implement support for deterministic-only macros?

**Comment by asv1** [comment]
Wondering, how the problem can be solved? Running external binary which run proc macros and collecting results? Or it can't be solved by intellij rust?

**Comment by vlad20012** [comment]
@asv1 yes, a kind of. We're going to support them. The thing that is sorely lacking is some way to make Cargo build all proc macros for us. But we can do without it - by building full project

**Comment by asv1** [comment]
@vlad20012 
But after building full project, there are no generated sources? Just curious, how it can be parsed.

**Comment by vlad20012** [comment]
Oh, it's a long story. If we have compiled proc macro `.so` (`.dll`) files, we can link and ask them to expand some macro. This approach is used in https://github.com/fedochet/rust-proc-macro-expander project (that was developed during JetBrains internship in 2018). Today the same approach is used in rust-analyzer, so usage of some rust-analyzer part for proc macro expansion is also considerable (see https://github.com/intellij-rust/intellij-rust/issues/964#issuecomment-627987541)

**Comment by Kobzol** [comment]
@vlad20012 just out of curiosity, is proc-macro support an ongoing effort (where outside contributors can maybe help?) or is it in a waiting phase, blocked by some upstream support that is not available yet?

**Comment by vlad20012** [comment]
It's something in the middle. It can be done right now, but can't be released (enabled by default) without 2 things. This is very similar to `build.rs` support. We implemented it in #1908, but it's still disabled. There are 2 reasons:

1. There is no way to ask `Cargo` to build _and run_ all `build.rs` scripts in the project (as well as build all proc macros). There is an issue https://github.com/rust-lang/cargo/issues/7178. Without this, we should rely on `cargo check` that is slow and unreliable (because of possible compilation errors in our workspace itself)
2. `Cargo Sync` task. Today we use `cargo metadata` command to get info about the workspace. We use it very extensively because it's just fast. But in the future, we will also need to run something like `cargo build --compile-time-deps` (or `cargo check` if https://github.com/rust-lang/cargo/issues/7178 is not implemented). This is significantly slower, so we need a separate `Sync` (or `Import`) task that looks just like `Gradle Sync` task (with a separate toolwindow, project refresh button and so on). @Undin is now working on this.

The same blockers are also applied to the proc macros implementation. Proc macro support itself doesn't look too complex btw.

If _someone_ wants to help, they can take a look at https://github.com/rust-lang/cargo/issues/7178 in Cargo. The issue mentions only build scripts, but really the command should also build all proc macro crates in the project.

**Comment by SCLeoX** [comment]
I have a question. How are people writing rust without this, at all? Call me spoiled but how do you even write code without autocomplete? Is there some other IDEs or editors that support this?

I am legitimately curious. Do people actually go back and forth between the code and the documentation to see what methods exist? If so, that sounds so unproductive.

Due to this, I have been trying to avoid procedure macros as much as possible, but that does not sound like a good way to go at all.

**Comment by Kobzol** [comment]
I don't think it's THAT common to have procedural macros that actually generate methods that you would actually use manually. It surely happens (for example crates like `getset`), but in such cases you can just lookup the methods, the docs are pretty great.

In my case, I mostly use procedural macros for things like `serde`, `thiserror`, `structopt` (these and similar crates are pretty popular). In these cases, the procedural macro only implements some trait that interacts with the rest of the ecosystem (`Serializable`, `Error`, etc.) and I don't have to call any of the implemented methods, so I don't even miss any autocomplete functionality.

You mentioned `diesel` - for example its `#[derive(Insertable)]` implements some trait that makes it possible to insert the struct into the database. When you want to insert the struct, you can create a struct literal, for that you don't have to call any (auto-generated) methods on the struct. Of course there are also some methods for example `Queryable` that won't be autocompleted and that are actually used often, but these are the same for all of the structs, so they should be easy to remember.

So yeah, the absence of autocomplete sucks, but I don't think that it's such a big problem.

**Comment by bes** [comment]
When I was a beginner Rustacean, 12 months ago, I was intrigued by the possibility to leave memory hogs such as Spring/Java and replace them with seemingly nice alternatives like Rocket/Diesel.

At that stage I found Rocket's documentation to be very good, and Diesel's getting started guide was at least good enough to get started. At least in combination with some random blog posts.

Then I started to flesh out my app with more and more SQL tables, and frankly I hit a brick wall.

At that stage, I was desperately seeking information on how to do really _basic_ things with Diesel, but the Diesel web page only covers the super basic things, and the crate documentation is literally impenetrable it's just pages of traits and generic parameters. I still can't write a good generic function to do things that are really easy to understand using e.g. jOOQ in Java, but that's besides the point.

It would have saved me a _lot_ of time to have that autocomplete available for things that Diesel generated in procedural macros.

I finally figured most of the things out by combining information from blogs, stackoverflow, and random github issues.

If I didn't really really like rust (which I do!) I probably would have just given up.

TL;DR
For beginner/intermediate rustacean procedural macro autocomplete is really important. As your understanding of the language and tooling (and `cargo expand`) grows, it becomes less important.

That's why I personally think it would be a great investment to have in intellij-rust. But I also understand that it's complicated and hard to implement.

**Comment by Kobzol** [comment]
Good point, the importance of procedural macro autocompletion is probably way more important for beginners, because the procedural system is *black magic* by itself and without autocomplete it can feel pretty hard to deal with in some cases.

**Comment by SCLeoX** [comment]
@bes My experience has been very similar, except my love for rust, unfortunately, was not strong enough and did not overcome this obstacle, and I just gave up.

**Comment by Restioson** [comment]
I'd say that upwards of 98% of crates that I use autocomplete. This is because most do not use proc macros in the public API. There is kind of basic proc macro support for impl blocks, which covers async_trait for instance - intellij assumes that it emits a method that is very similar to what is defined, and it autocompletes like that (or at least it feels like that).

**Comment by NyxCode** [comment]
When I was starting with rust, I too really missed autocomplete for generated code. However, over the years, I found that just using rustdocs is enough most of the time. Especially since derive macros just implement a trait, you can read the documentation of that trait.
I see that many people are struggeling with diesel, and I get why. I am using it in a fairly big project, and I *really* regret chosing it. I think it's an interesting experiment, but utterly impractical. Working with diesel gets so verbose, that I don't see how it's better than just writing raw sql. The approach sqlx has taken seems far superior, and I would recommend it over diesel 100%.

**Comment by panstromek** [comment]
It's similar experience with scripting languages. When I went from Java to PHP and JavaScript, it felt like a major downgrade in IDE experience because those languages are too dynamic and IDE can't analyze them that well. You get used to it, but there are also ways to hack around it get at least some help from the IDE. In JavaScript and PHP, you can put types in jsdoc/phpdoc comments and IDE will resolve them and autocomplete.

```js
/** @type {HTMLCanvasElement} */
const canvas = document.getElementById('canvas');
```

Similarly, in Rust, you can give type hints that helps the IDE, too. For example:

```rust
pub trait MyTrait {
    fn trait_method(&self, i :i32) -> i64;
}
#[derive(MyTrait)]
struct MyStruct;
fn trampoline(i : MyStruct) -> impl MyTrait { i }

fn foo() {
    let i = MyStruct;
    // let i = trampoline(i); // if you uncomment this line, IDE will resolve the `trait_method`
    let x = i.trait_method();
}
```

Even simpler solution is to use `impl Trait` binding:

```rust
fn foo() {
    let i : impl MyTrait = MyStruct;
    let x = i.trait_method();
}
```
This syntax is unstable, though, so it's only good to add temporarily while you write code.


If you have a generated function and want autocomplete for return type, you can annotate the call with incorrect type and run `cargo check`. Compiler will tell you the correct type and you can copypaste it in there.

```rust
fn foo() {
    let gen : u64 = returns_something_else();
}
```

This is also helpful with functions that return collections of generated types, like `Vec<GeneratedStruct>`. You can annotate it like this and you'll get at least partial autocomplete for `Vec` methods.

```rust
fn foo() {
    let gen : Vec<_> = returns_something_else(); // returns Vec<GeneratedStruct>
}
```

One special case is `web_sys` crate and other `wasm_bindgen` generated code, where functions exist and IDE can see them, but they get transormed by the bindgen such that they end up in different namespace with a bit different arguments. So you can take advantage of some semi-correct autocomplete, build some code that doesn't actually compile and change it to the correct syntax once you actually  need to compile and run it. Not ideal, but it can help.

**Comment by changhe3** [comment]
Also almost all parser generators use procedural macro to generate code. It's quite annoying to use them without autocompleting.

**Comment by Pscheidl** [comment]
Same goes for Kubernetes - the allmighty [Kube-runtime](https://crates.io/crates/kube-runtime) crate basically stands and falls with generating new structs. Using those without autocomplete is just hell.

**Comment by vultix** [comment]
I'd love to have this for [mockall](https://docs.rs/mockall/0.8.3/mockall/)

**Comment by arn-the-long-beard** [comment]
> Also almost all parser generators use procedural macro to generate code. It's quite annoying to use them without autocompleting.

I agree and it is so needed. Is there a way we could help to make the autocompleting happens ?

**Comment by Undin** [comment]
@arn-the-long-beard we have already started to implement the corresponding support. See https://github.com/intellij-rust/intellij-rust/pull/6564 as the first user-visible step.
Eventually, it's supposed to work in a similar way as the current declarative macro expansion work at this moment

**Comment by arn-the-long-beard** [comment]
Okay ! 
Thank you so much for your message !

I wait then :smile: .

Keep up the good work :wink:

**Comment by bes** [comment]
I'm not sure if this idea has been floated, or maybe if it's super obvious but here goes:

Proc macros (and declarative macros) have quite poor discoverability. What if intellij-rust created a virtual directory, like `macro-generated` in the directory structure, where the generated code can be read like any other source file (but not edited) and navigated to just like any other piece of code.

Perhaps a generated directory isn't 100% correct since macros expand "into" the source, so maybe something like the new "rendered documentation" feature of intellij would be better.

I think this could be a quite unique selling feature of intellij-rust/Clion.

TL;DR
Don't just make macro code discoverable through autocomplete, make the source readable/navigatable as well!

Please let me know if this should spin off into a separate issue! @Undin

**Comment by panstromek** [comment]
@bes Well, if I underestand correctly, you can already do that if you choose `alt+enter | show (recursive) macro expansion` on some macro item. I use this quite a lot, it's handy. It's obviously not possible for proc macros now, because plugin doesn't support them, yet, but I assume it'll be supported for them in the future, too.

**Comment by bes** [comment]
> @bes Well, if I underestand correctly, you can already do that if you choose `alt+enter | show (recursive) macro expansion` on some macro item. I use this quite a lot, it's handy. It's obviously not possible for proc macros now, because plugin doesn't support them, yet, but I assume it'll be supported for them in the future, too.

Yes, but it's a difference between an action to generate something on demand (for proc macros there's [cargo expand](https://github.com/dtolnay/cargo-expand)), and having them available ambiently.

I think having them available ambiently, and navigatable, is a game changer for macros...

**Comment by Undin** [comment]
@bes @panstromek 

> Proc macros (and declarative macros) have quite poor discoverability. What if intellij-rust created a virtual directory, like macro-generated in the directory structure, where the generated code can be read like any other source file (but not edited) and navigated to just like any other piece of code.
>
> Perhaps a generated directory isn't 100% correct since macros expand "into" the source, so maybe something like the new "rendered documentation" feature of intellij would be better.
>
> I think this could be a quite unique selling feature of intellij-rust/Clion.
>
> TL;DR
> Don't just make macro code discoverable through autocomplete, make the source readable/navigatable as well!

Actually, the plugin already does such things (almost). I.e. there is a virtual file system where the plugin keeps expanded macros but each macro expansion step has own file. It allows us to reuse existing capabilities of IntelliJ platform and provide expected code insight features for/w generated items like name resolution, completion, etc. But currently, we intentionally hide these virtual files not to confuse users with a LOT of intermediate files.

So, the question - what exactly do you want to do with these expansion files?

>  It's obviously not possible for proc macros now, because plugin doesn't support them, yet, but I assume it'll be supported for them in the future, too.

Yep, #6564 uses the same approach to keep results of proc macro expansion so `show (recursive) macro expansion` actions work as expected

> Please let me know if this should spin off into a separate issue!

It would be great to have a separate issue for the desired feature to collect all necessary details in one place

**Comment by bes** [comment]
I have forked the discussion about "virtual folders" to #6875.

**Comment by mili-l** [comment]
see #9729
